social_core/backends/apple.py

```diff
@@ -0,0 +1,137 @@
+    * `SECRET` - your secret key;
+    * `SCOPE` (optional) - e.g. `['name', 'email']`;
+    * `EMAIL_AS_USERNAME` - use apple email is username is set, use apple id otherwise.
+    * `AppleIdAuth.TOKEN_TTL_SEC` - time before JWT token expiration, seconds.
+"""
+
+import json
+import time
+
+import jwt
+from jwt.algorithms import RSAAlgorithm
+from jwt.exceptions import PyJWTError
+
+from social_core.backends.oauth import BaseOAuth2
+from social_core.exceptions import AuthCanceled
+
+
+class AppleIdAuth(BaseOAuth2):
+    name = "apple-id"
+
+    JWK_URL = "https://appleid.apple.com/auth/keys"
+    AUTHORIZATION_URL = "https://appleid.apple.com/auth/authorize"
+    ACCESS_TOKEN_URL = "https://appleid.apple.com/auth/token"
+    ACCESS_TOKEN_METHOD = "POST"
+    RESPONSE_MODE = None
+
+    ID_KEY = "sub"
+    TOKEN_KEY = "id_token"
+    STATE_PARAMETER = True
+    REDIRECT_STATE = False
+
+    TOKEN_AUDIENCE = "https://appleid.apple.com"
+    TOKEN_TTL_SEC = 6 * 30 * 24 * 60 * 60
+
+    def auth_params(self, *args, **kwargs):
+        """Apple requires to set `response_mode` to `form_post` if `scope` parameter is passed."""
+        params = super(AppleIdAuth, self).auth_params(*args, **kwargs)
+        if self.RESPONSE_MODE:
+            params["response_mode"] = self.RESPONSE_MODE
+        elif self.get_scope():
+            params["response_mode"] = "form_post"
+        return params
+
+    def get_private_key(self):
+        """
+        Return contents of the private key file.
+        Override this method to provide secret key from another source if needed.
+        """
+        return self.setting("SECRET")
+
+    def generate_client_secret(self):
+        now = int(time.time())
+        client_id = self.setting("CLIENT")
+        team_id = self.setting("TEAM")
+        key_id = self.setting("KEY")
+        private_key = self.get_private_key()
+
+        headers = {"kid": key_id}
+        payload = {
+            "iss": team_id,
+            "iat": now,
+            "exp": now + self.TOKEN_TTL_SEC,
+            "aud": self.TOKEN_AUDIENCE,
+            "sub": client_id,
+        }
+
+        return jwt.encode(payload, key=private_key, algorithm="ES256", headers=headers)
+
+    def get_key_and_secret(self):
+        client_id = self.setting("CLIENT")
+        client_secret = self.generate_client_secret()
+        return client_id, client_secret
+
+    def get_apple_jwk(self):
+        """Returns an iterable of apple jwk strings."""
+        keys = self.get_json(url=self.JWK_URL).get("keys")
+        if not isinstance(keys, list) or not keys:
+            raise AuthCanceled("Invalid jwk response")
+
+        return (json.dumps(key) for key in keys)
+
+    def decode_id_token(self, id_token):
+        """Decode and validate JWT token from apple and return payload including user data."""
+        if not id_token:
+            raise AuthCanceled("Missing id_token parameter")
+
+        for jwk_string in self.get_apple_jwk():
```

Comment:
So I've been following this PR for a few months and know this iterating retry logic was not here originally. Out of curiosity, what led to this change? Did something change on Apple's side?

Reply:
Yesterday Apple started to sign with two new certs. See https://appleid.apple.com/auth/keys

Iterate is not the best approach, but will be great if this PR can be merged, I've two improvements ready to be proposed